lib/ansible/modules/notification/mattermost.py

```diff
@@ -0,0 +1,158 @@
+    default: None
+  username:
+    description:
+      - This is the sender of the message (Username Override need to be enabled by mattermost admin, see mattermost doc, default C(Ansible)).
+    required: false
+    default: "Ansible"
+  icon_url:
+    description:
+      - Url for the message sender's icon (default C(https://www.ansible.com/favicon.ico))
+    required: false
+  validate_certs:
+    description:
+      - If C(no), SSL certificates will not be validated. This should only be used
+        on personally controlled sites using self-signed certificates.
+    required: false
+    default: 'yes'
+    choices:
+      - 'yes'
+      - 'no'
+"""
+
+EXAMPLES = """
+- name: Send notification message via Mattermost
+  mattermost:
+    url: http://mattermost.example.com
+    api_key: my_api_key
+    text: '{{ inventory_hostname }} completed'
+
+- name: Send notification message via Slack all options
+  mattermost:
+    url: http://mattermost.example.com
+    api_key: my_api_key
+    text: '{{ inventory_hostname }} completed'
+    channel: notifications
+    username: 'Ansible on {{ inventory_hostname }}'
+    icon_url: http://www.example.com/some-image-file.png
+"""
+
+RETURN = '''
+payload:
+    description: Mattermost payload
+    returned: success
+    type: string
+webhook_url:
+    description: URL the webhook is sent to
+    returned: success
+    type: string
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.urls import fetch_url
+
+def main():
+    module = AnsibleModule(
+        supports_check_mode=True,
+        argument_spec = dict(
+            url         = dict(type='str', required=True),
+            api_key     = dict(type='str', required=True,  no_log=True),
+            text        = dict(type='str', required=True),
+            channel     = dict(type='str', default=None),
+            username    = dict(type='str', default='Ansible'),
+            icon_url    = dict(type='str', default='https://www.ansible.com/favicon.ico'),
+            validate_certs = dict(default='yes', type='bool'),
+        )
+    )
+    #init return dict
+    result = dict(changed=False, msg="OK")
+
+    #define webhook
+    webhook_url = "{0}/hooks/{1}".format(module.params['url'],module.params['api_key'])
+    result['webhook_url'] = webhook_url
+
+    #define payload
+    payload = { }
+    for param in ['text', 'channel', 'username', 'icon_url']:
+        if module.params[param] is not None:
+            payload[param] = module.params[param]
+
+    payload=module.jsonify(payload)
+    result['payload'] = payload
+
+    #http headers
+    headers = {
+        'Content-Type': 'application/json',
+        'Accept': 'application/json',
+    }
+
+    #send request if not in test mode
+    if module.check_mode is False:
+        response, info = fetch_url(module=module, url=webhook_url, headers=headers, method='POST', data=payload)
```

Comment:
Right, but now check_mode is not testing anything at all. We might as well have no check_mode in this code.
I was hoping to see an implementation that looks like this: https://github.com/ansible/ansible/pull/19348/files#diff-473c3ae27a58aa5098bfcf29652164d8R141

So the solution implemented here was doing only the authentication to check connectivity and credentials (so it properly fails as if it would be running properly). Not sure if this is possible, but that would be as close as possible to the real thing.